tests/doc_test_framework.go

```diff
@@ -0,0 +1,255 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package tests
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"testing"
+
+	"istio.io/istio.io/pkg/test/istioio"
+	"istio.io/istio/pkg/test/framework"
+)
+
+// TestCase is a description of a test extracted from a file
+type TestCase struct {
+	valid bool  // whether this is a valid test case that can be run
+	err   error // whether there is an error when constructing the test case
+
+	path          string // path of the test file
+	config        string // setup config of the test
+	testScript    string // test script to be run
+	cleanupScript string // cleanup script to be run
+}
+
+var (
+	testsToRun   = os.Getenv("TEST")
+	runAllTests  = (testsToRun == "")
+	testsAsSlice = split(testsToRun)
+
+	// folder location to be traversed to look for test files
+	contentFolder = fmt.Sprintf("%v/content/", os.Getenv("REPO_ROOT"))
+
+	// scripts that are sourced for all tests
+	helperTemplate = `
+		cd ${REPO_ROOT}
+		source "content/%v" # snips.sh
+		source "tests/util/verify.sh"
+		source "tests/util/debug.sh"
+		source "tests/util/helpers.sh"
+	`
+
+	snipsFileSuffix = "/snips.sh"
+	testFileSuffix  = "/test.sh"
+
+	setupSpec      = "# @setup"
+	testCleanupSep = "# @cleanup"
+
+	// constructed test cases for the specified tests
+	testCases []TestCase
+)
+
+// init constructs test cases from all specified tests
+func init() {
+	if runAllTests {
+		log.Println("Starting test doc(s): all docs")
+	} else {
+		log.Println("Starting test doc(s):", testsToRun)
+	}
+
+	// scan the test script files
+	err := filepath.Walk(
+		contentFolder,
+		func(path string, info os.FileInfo, walkError error) error {
+			if walkError != nil {
+				return walkError
+			}
+			if testCase := checkFile(path); testCase.valid {
+				testCases = append(testCases, *testCase)
+			} else if testCase.err != nil {
+				log.Printf("Error occurred while processing %v: %v", testCase.path, testCase.err)
+			}
+			return nil
+		},
+	)
+	if err != nil {
+		log.Fatalln("Error occurred while traversing content:", err)
+	}
+
+	// in case no matched script files were found
+	if len(testCases) == 0 {
+		log.Printf("Warning: no test scripts are found that match '%v'", testsToRun)
+	}
+}
+
+// checkFile takes a file path as the input and returns a TestCase object
+// that is constructed out of the file as a file description
+func checkFile(path string) *TestCase {
+	// check if ends with test.sh and if required to run
+	if !strings.HasSuffix(path, testFileSuffix) {
+		return &TestCase{}
+	} else if !runAllTests && !matched(path, testsAsSlice) {
+		return &TestCase{}
+	}
+	shortPath := path[len(contentFolder):]
+
+	// read the script file
+	script, err := ioutil.ReadFile(path)
+	if err != nil {
+		return &TestCase{path: shortPath, err: err}
```

Comment:
Usually in go, the error is returned as a second argument in the function instead of as a property of the returned object

Reply:
Gotcha. Changed in the commit below.